ocaml/xapi/message_forwarding.ml

```diff
@@ -5287,6 +5287,36 @@ functor
         info "SDN_controller.forget: sdn_controller = '%s'"
           (sdn_controller_uuid ~__context self) ;
         Local.SDN_controller.forget ~__context ~self
+
+      let open_port ~__context ~self ~protocol ~port =
+        info
+          "SDN_controller.open_port: sdn_controller = '%s', protocol = '%s', \
+           port = '%s'"
+          (sdn_controller_uuid ~__context self)
+          (Record_util.sdn_port_protocol_to_string protocol)
+          (Int64.to_string port) ;
+        let local_fn = Local.SDN_controller.open_port ~self ~protocol ~port in
+        List.iter
+          (fun host ->
+            do_op_on ~local_fn ~__context ~host (fun session_id rpc ->
+                Client.SDN_controller.open_port rpc session_id self protocol
+                  port))
+          (Db.Host.get_all ~__context)
+
+      let close_port ~__context ~self ~protocol ~port =
+        info
+          "SDN_controller.close_port: sdn_controller = '%s', protocol = '%s', \
+           port = '%s'"
+          (sdn_controller_uuid ~__context self)
+          (Record_util.sdn_port_protocol_to_string protocol)
+          (Int64.to_string port) ;
+        let local_fn = Local.SDN_controller.close_port ~self ~protocol ~port in
+        List.iter
+          (fun host ->
+            do_op_on ~local_fn ~__context ~host (fun session_id rpc ->
+                Client.SDN_controller.close_port rpc session_id self protocol
+                  port))
+          (Db.Host.get_all ~__context)
```

Comment:
Have you tested that both of these functions are idempotent?

For example, suppose you run `SDN_controller.close_port`, and it succeed on host A,B, but fails on C. Then the user will probably want to run `SDN_controller.close_port` again once they have fixed the problem on C - will that work?

(In addition we probably want to have the loops closing/opening ports to close/open as many ports as possible, and then raise any exceptions later)

Reply:
Yes it'd work, you can call several times `open_port`/`close_port` without any problem, even if the port is already opened/closed.

So you think i should catch the error in the loop and throw it after the loop completion? What if several iterations fail, should i throw as many error?